crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,582 @@
+            trace!("header_by_num: got none");
+            return Ok(None);
+        }
+
+        trace!("header_by_num: got");
+
+        Ok(block.map(|b| b.header.inner))
+    }
+
+    /// Gets block number of latest known block
+    fn last_block_number(&self) -> ProviderResult<BlockNumber> {
+        let id: u64 = rand::random();
+        let span = trace_span!("last_block_num", id);
+        let _guard = span.enter();
+        trace!("last_block_num:get");
+
+        let block_num = self
+            .ipc_provider
+            .call::<_, U64>("eth_blockNumber", ())
+            .map_err(ipc_to_provider_error)?
+            .to::<u64>();
+
+        trace!("last_block_num: got");
+        Ok(block_num)
+    }
+
+    /// Creates new root hasher - struct responsible for calculating root hash
+    fn root_hasher(&self, parent_hash: BlockNumHash) -> ProviderResult<Box<dyn RootHasher>> {
+        Ok(Box::new(StatRootHashCalculator {
+            remote_provider: self.ipc_provider.clone(),
+            parent_hash: parent_hash.hash,
+        }))
+    }
+}
+
+#[derive(Clone)]
+pub struct IpcStateProvider {
+    ipc_provider: RpcProvider,
+    block_id: BlockId,
+
+    //Per block cache
+    block_hash_cache: DashMap<u64, BlockHash>,
+    storage_cache: DashMap<(Address, StorageKey), StorageValue>,
+    account_cache: DashMap<Address, Account>,
+
+    // Global cache (cache not related to specific block)
+    code_cache: Arc<DashMap<B256, Bytecode>>,
+}
+
+impl IpcStateProvider {
+    /// Crates new instance of state provider
+    fn new(
+        ipc_provider: RpcProvider,
+        block_id: BlockId,
+        code_cache: Arc<DashMap<B256, Bytecode>>,
+    ) -> Self {
+        Self {
+            ipc_provider,
+            block_id,
+
+            code_cache,
+
+            block_hash_cache: DashMap::new(),
+            storage_cache: DashMap::new(),
+            account_cache: DashMap::new(),
+        }
+    }
+
+    /// Crates new instance of state provider on the heap
+    // Box::new(Arc::new(Self)) is required because StateProviderFactory returns Box<dyn StateProvider>
+    #[allow(clippy::redundant_allocation)]
+    fn into_boxed(
+        ipc_provider: RpcProvider,
+        block_id: BlockId,
+        code_cache: Arc<DashMap<B256, Bytecode>>,
+    ) -> Box<Arc<Self>> {
+        Box::new(Arc::new(Self::new(ipc_provider, block_id, code_cache)))
+    }
+}
+
+impl StateProvider for IpcStateProvider {
+    /// Get storage of given account
+    fn storage(
+        &self,
+        account: Address,
+        storage_key: StorageKey,
+    ) -> ProviderResult<Option<StorageValue>> {
+        let id: u64 = rand::random();
+        let span = trace_span!("storage", id);
+        let _guard = span.enter();
+        trace!("storage:get");
+
+        if let Some(storage) = self.storage_cache.get(&(account, storage_key)) {
+            return Ok(Some(*storage));
+        }
+
+        let key: U256 = storage_key.into();
+        let storage = self
```

Comment:
Could it return None?

Reply:
I see this in reIpc:
```rust
            ResponsePayload::Success(_) => {
                match resp.try_success_as::<T>() {
                    Some(Ok(r)) => Ok(r),
                    Some(Err(e)) => Err(e.into()),
                    // the response was received successfully, but it contains Err payload
                    // we shouldn't have  ended up here
                    None => Err(RpcError::JsonErrPayloadMisinterpretedAsSuccess),
                }
            }
```

Reply:
So i assume you return an error on empty response

Reply:
We discussed this during call, but I'll also leave the comment here. 

I'm not handling the `None` case in all of these calls you pointed out because I wrote the API on the Nethermind client so that it will either succeed and return some value or return an error. 
Meaning there isn't any scenario in which the response from the server would return smth. like `Ok<None>`.
But I agree with your comments, this was bad design, which has now been fixed :)

------

Regarding the comment about reipc, the Error is not returned on empty response. 
The `None` branch in code above is hit in case Response was marked as Success, but during deserialisation it was deserialised as Error - I'm not sure in which scenario this happens; I don't think it should ever happen.